src/simulators/statevector/kernel_simd.hpp

```diff
@@ -0,0 +1,379 @@
+
+#ifndef _mm256_set_m128d
+#define _mm256_set_m128d(hi,lo) _mm256_insertf128_pd(_mm256_castpd128_pd256(lo), (hi), 0x1)
+#endif
+#ifndef _mm256_storeu2_m128d
+#define _mm256_storeu2_m128d(hiaddr,loaddr,a) \
+        do { __m256d _a = (a); _mm_storeu_pd((loaddr), \
+             _mm256_castpd256_pd128(_a)); \
+             _mm_storeu_pd((hiaddr), \
+             _mm256_extractf128_pd(_a, 0x1));\
+            } while (0)
+#endif
+#ifndef _mm256_loadu2_m128d
+#define _mm256_loadu2_m128d(hiaddr,loaddr) _mm256_set_m128d(_mm_loadu_pd(hiaddr), _mm_loadu_pd(loaddr))
+#endif
+
+const __m256d neg = _mm256_setr_pd(1.0, -1.0, 1.0, -1.0);
+
+inline __m256d mul(__m256d const& c1, __m256d const& c2, __m256d const& c2_perm){
+    auto acbd = _mm256_mul_pd(c1, c2);
+    auto acbd_perm = _mm256_mul_pd(c1, c2_perm);
+    return _mm256_hsub_pd(acbd, acbd_perm);
+}
+
+inline __m256d loadu2(double const*p1, double const*p2){
+    return _mm256_loadu2_m128d((double const*)p2, (double const*)p1);
+}
+
+inline __m256d loadbc(double *p){
+    auto tmp = _mm_load_pd((const double*)p);
+    return _mm256_broadcast_pd(&tmp);
+}
+
+inline void kernel1_init(__m256d *mat_vec, __m256d *mat_perm, const cvector_t<double> &mat){ 
+    mat_vec[0] = loadu2((double *)&mat[0], (double *)&mat[1]); 
+    mat_vec[1] = loadu2((double *)&mat[2], (double *)&mat[3]); 
+
+    for (unsigned i = 0; i < 2; ++i){
+      auto badc = _mm256_permute_pd(mat_vec[i], 5);
+      mat_perm[i] = _mm256_mul_pd(badc, neg);
+    }
+}
+
+template<typename data_t=double>
+inline void kernel1(std::complex<data_t> *data_, const areg_t<2> &inds, __m256d *mat_vec, __m256d *mat_perm ){
+    __m256d vec[2];
+    vec[0] = loadbc((double *)&data_[inds[0]]);
+    vec[1] = loadbc((double *)&data_[inds[1]]);
+
+
+    _mm256_storeu2_m128d((double *)&data_[inds[1]], (double *)&data_[inds[0]], 
+        _mm256_add_pd(mul(vec[0], mat_vec[0], mat_perm[0]), mul(vec[1], mat_vec[1], mat_perm[1])));
+}
+
+inline void kernel2_init(__m256d *mat_vec, __m256d *mat_perm, const cvector_t<double> &mat){ 
+
+    mat_vec[0]  = loadu2((double *)&mat[0], (double *)&mat[1]);
+    mat_vec[1]  = loadu2((double *)&mat[4], (double *)&mat[5]);
+    mat_vec[2]  = loadu2((double *)&mat[8], (double *)&mat[9]);
+    mat_vec[3]  = loadu2((double *)&mat[12], (double *)&mat[13]);
+    mat_vec[4]  = loadu2((double *)&mat[2], (double *)&mat[3]);
+    mat_vec[5]  = loadu2((double *)&mat[6], (double *)&mat[7]);
+    mat_vec[6]  = loadu2((double *)&mat[10], (double *)&mat[11]);
+    mat_vec[7]  = loadu2((double *)&mat[14], (double *)&mat[15]);
+
+    for (unsigned i = 0; i < 8; ++i){
+      auto badc = _mm256_permute_pd(mat_vec[i], 5);
+      mat_perm[i] = _mm256_mul_pd(badc, neg);
+    }
+}
+
+template<typename data_t=double>
+inline void kernel2(std::complex<data_t> *data_, const areg_t<4> &inds, __m256d *mat_vec, __m256d *mat_perm ){
+    __m256d vec[4];
+
+    vec[0] = loadbc((double *)&data_[inds[0]]);
```

Comment:
Is there a specific reason you are only using half available space in `__m256d` (If I understand correctly you are storing 2 doubles when they can hold 4 doubles)

Reply:
This call stores 4 doubles. `loadbc` calls `_mm_load_pd` and `_mm256_broadcast_pd` so vec[0] has inds[0] and inds[0]. SIMD uses this 2 x inds[0] varibales.

Reply:
Ah I see!